File: src/daybreak/adapters/terminal/jsonc.py
import json
import re
from pathlib import Path


def load_jsonc_file(path: Path):
    raw_text = path.read_text(encoding="utf-8")
    return parse_jsonc(raw_text)


def dump_json_file(path: Path, data: dict):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, indent=4, ensure_ascii=False) + "\n", encoding="utf-8")


def parse_jsonc(text: str):
    stripped = _strip_comments(text)
    stripped = _remove_trailing_commas(stripped)
    return json.loads(stripped)


def _strip_comments(text: str) -> str:
    result = []
    in_string = False
    in_line_comment = False
    in_block_comment = False
    escaped = False
    i = 0
    length = len(text)

    while i < length:
        ch = text[i]
        nxt = text[i + 1] if i + 1 < length else ""

        if in_line_comment:
            if ch == "\n":
                in_line_comment = False
                result.append(ch)
            i += 1
            continue

        if in_block_comment:
            if ch == "*" and nxt == "/":
                in_block_comment = False
                i += 2
                continue
            if ch == "\n":
                result.append(ch)
            i += 1
            continue

        if in_string:
            result.append(ch)
            if escaped:
                escaped = False
            elif ch == "\\":
                escaped = True
            elif ch == '"':
                in_string = False
            i += 1
            continue

        if ch == "/" and nxt == "/":
            in_line_comment = True
            i += 2
            continue
        if ch == "/" and nxt == "*":
            in_block_comment = True
            i += 2
            continue
        if ch == '"':
            in_string = True

        result.append(ch)
        i += 1

    return "".join(result)


def _remove_trailing_commas(text: str) -> str:
    pattern = re.compile(r",(\s*[}\]])")
    previous = None
    current = text
    while previous != current:
        previous = current
        current = pattern.sub(r"\1", current)
    return current

File: src/adapters/terminal/jsonc.ts
import * as fs from 'fs';
import * as path from 'path';

export function loadJsoncFile(filePath: string): any {
  const raw = fs.readFileSync(filePath, 'utf-8');
  return parseJsonc(raw);
}

export function dumpJsonFile(filePath: string, data: any): void {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(data, null, 4) + '\n', 'utf-8');
}

export function parseJsonc(text: string): any {
  const stripped = removeTrailingCommas(stripComments(text));
  return JSON.parse(stripped);
}

function stripComments(text: string): string {
  const result: string[] = [];
  let inString = false;
  let inLineComment = false;
  let inBlockComment = false;
  let escaped = false;
  let i = 0;
  const length = text.length;

  while (i < length) {
    const ch = text[i];
    const nxt = i + 1 < length ? text[i + 1] : '';

    if (inLineComment) {
      if (ch === '\n') { inLineComment = false; result.push(ch); }
      i++;
      continue;
    }
    if (inBlockComment) {
      if (ch === '*' && nxt === '/') { inBlockComment = false; i += 2; continue; }
      if (ch === '\n') result.push(ch);
      i++;
      continue;
    }
    if (inString) {
      result.push(ch);
      if (escaped) {
        escaped = false;
      } else if (ch === '\\') {
        escaped = true;
      } else if (ch === '"') {
        inString = false;
      }
      i++;
      continue;
    }
    if (ch === '/' && nxt === '/') { inLineComment = true; i += 2; continue; }
    if (ch === '/' && nxt === '*') { inBlockComment = true; i += 2; continue; }
    if (ch === '"') inString = true;
    result.push(ch);
    i++;
  }

  return result.join('');
}

function removeTrailingCommas(text: string): string {
  const pattern = /,(\s*[}\]])/g;
  let prev: string | null = null;
  let current = text;
  while (prev !== current) {
    prev = current;
    current = current.replace(pattern, '$1');
  }
  return current;
}
